### PR TITLE
extended connector  with subprocesses and functions for downloading f…

### PR DESCRIPTION
**Pull request by ramazanakbulut**
Hi reguel, 
can you please review my changes for the docuware connecter (+ demo project)
as we discussed a few days ago i added some sub processes and functions mainly for uploading files inculding index fields (meta data) and for updating these index fields

Many Thanks 
Ramazan

### CODE REVIEW

File: docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java [hunk 1]
```diff
@@ -55,4 +95,139 @@ public Date stringToDate(String dateString) {
...
+		FileUtils.write(propertiesFile.getJavaFile(), JsonUtils.serializeProperties(docuWareproperties), PROPERTIES_FILE_CHARSET);
+		return propertiesFile;
+	}
+	
+	private static Builder prepareRestClient(WebTarget target, DocuWareEndpointConfiguration configuration) {
+		return target.request().header("X-Requested-By", "ivy")
+				.header("MIME-Version", "1.0")
+				.header("Accept", "application/xml")
+				.header("Connection", "keep-alive")
+				.header("Authorization", AUTHORIZATION_BASIC + configuration.getCredentials());
```
Comment (ivy-rew) [inline docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java:192]: authorization should already be solved by the DocuwareAuthFeature. can you outline why this additional basic auth is required?

File: docuware-connector/src/com/axonivy/connector/docuware/connector/utils/JsonUtils.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+import ch.ivyteam.ivy.environment.Ivy;
+import ch.ivyteam.util.StringUtil;
+
+/**
+ * Provides utilities related to json serialization like building an ObjectMapper with all required options and features.
+ *
+ * @author jpl
+ *
+ */
+public class JsonUtils {
```
Comment (ivy-rew) [inline docuware-connector/src/com/axonivy/connector/docuware/connector/utils/JsonUtils.java:29]: heavy hammer that invents lot of code already present in the JAX-RS stack.
we only need it read/write this properites file right?

File: docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareEndpointConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package com.axonivy.connector.docuware.connector;
+
+public class DocuWareEndpointConfiguration {
+
+	
+	private String username;
+	
+	private String password;
```
Comment (ivy-rew) [inline docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareEndpointConfiguration.java:8]: secrets should not be held as String ... rather as byte[] to make heap-dump memory analysis harder. 
but hopefully we can get rid of this altogether, and prefer an authentication feature to deal with the user:pass

File: docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java [hunk 4]
```diff
@@ -55,4 +95,139 @@ public Date stringToDate(String dateString) {
...
+	}
+	
+	public static Document uploadStream(WebTarget target, byte[] file, String fileName, DocuWareEndpointConfiguration configuration, List<DocuWareProperty> properties) throws IOException, DocuWareException {
+		FormDataMultiPart multipart;
+		File propertiesFile = createPropertiesFile(properties);
+		
+		try (FormDataMultiPart formDataMultiPart = new FormDataMultiPart()) {		
+			InputStream streamProperties = new FileInputStream(propertiesFile.getJavaFile());
+			StreamDataBodyPart streamPropertiesPart = new StreamDataBodyPart(PROPERTIES_FILE_NAME, streamProperties );
+			streamPropertiesPart.setMediaType(MediaType.APPLICATION_JSON_TYPE);
```
Comment (ivy-rew) [inline docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java:111]: this is ok for now: but let's clarify why we need a special uploader today:
- [ ] the property-order of the form is important ... before 10.0.2 we could not guarantee sending order.
- [ ] the byte[] natured payload is easier to handle than files. 
is that correct or anything else involved?

**Review by ivy-rew — CHANGES_REQUESTED** [review]
Thanks a lot for contributing these wrappers for complex docuware calls back to the community. 

Some open questions from my site, but no blocker.

Can you add a simple web-test or bpmn-test to verify that the sub-process actually work? 

- [x] simple test

File: docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java (hunk 1 above)
Comment (ramazanakbulut) [reply docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java]: yes, thats true. i have removed this from the header (and also from the configuration helper class)

File: docuware-connector/src/com/axonivy/connector/docuware/connector/utils/JsonUtils.java (hunk 2 above)
Comment (ramazanakbulut) [reply docuware-connector/src/com/axonivy/connector/docuware/connector/utils/JsonUtils.java:22]: yes, i removed all unused methods

File: docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareEndpointConfiguration.java (hunk 3 above)
Comment (ramazanakbulut) [reply docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareEndpointConfiguration.java]: removed it togheter with the unnecessary additional basic auth in the header params

File: docuware-connector-test/src/com/axonivy/market/docuware/connector/test/xml/documentIndexFields.xml [hunk 5]
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" encoding="utf-8"?>
+<?xml-stylesheet type="text/xsl" href="/DocuWare/Platform/Content/standard.xslt"?>
+<DocumentIndexFields xmlns:s="http://dev.docuware.com/schema/public/services" xmlns="http://dev.docuware.com/schema/public/services/platform">
```
Comment (ivy-rew) [inline docuware-connector-test/src/com/axonivy/market/docuware/connector/test/xml/documentIndexFields.xml:3]: great: these mocks will help a lot to understand these index-fields better :+1:

**Review by ivy-rew — APPROVED** [review]

File: docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java (hunk 4 above)
Comment (ramazanakbulut) [reply docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java:116]: besides the issue with the property order and the bye[] payload one another reason was that if we want to use a json file for the document-index update we can not define the content-type (application/json) for this file (only for this index file) in the rest client call activity. Without the content-type information the upload did not work properly.

**Comment by ramazanakbulut** [comment]
Hi Reguel,  thank you for the review. i tried to fix the points. Besides that i added BPM Tests for the sub-processes

**Comment by ivy-rew** [comment]
> Hi Reguel, thank you for the review. i tried to fix the points. Besides that i added BPM Tests for the sub-processes

Thanks a lot @ramazanakbulut . For me the this change is good and well covered with tests.

File: docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java [hunk 6]
```diff
@@ -55,4 +93,126 @@ public Date stringToDate(String dateString) {
...
+	}
+	
+	public static Document uploadStream(WebTarget target, ch.ivyteam.ivy.scripting.objects.List<Byte> fileBytes, String fileName, DocuWareEndpointConfiguration configuration, List<DocuWareProperty> properties) throws IOException, DocuWareException {
+		Byte[] bytes = fileBytes.toArray(new Byte[fileBytes.size()]);
+		byte[]  byteArray =  ArrayUtils.toPrimitive(bytes);
+		
+		return uploadStream(target, byteArray,  fileName,  configuration,  properties);
+	}
+	
+	public static Document uploadStream(WebTarget target, byte[] file, String fileName, DocuWareEndpointConfiguration configuration, List<DocuWareProperty> properties) throws IOException, DocuWareException {
```
Comment (ivy-rew) [inline docuware-connector/src/com/axonivy/connector/docuware/connector/DocuWareService.java:109]: I hope we can soon provide 'inputStream/byte' support directly on the inscription-mask, and then get rid of this file-upload. 
Until then, this serves as great documentation for the circumstance. Thanks

**Comment by ivy-rew** [comment]
I merged it with :crossed_fingers:  can you verify whether this is a true error stated in the pipeline? 
or a false positive?

**Comment by ramazanakbulut** [comment]
ah sorry, i forgot to update this in the demo project. i just removed the wrong return parameter

**Comment by ivy-rew** [comment]
can you fire a fix PR for this @ramazanakbulut ? 
our build monitor is blaming this repo for its quality -> https://axonivy-market.github.io/market-monitor/

**Pull request merged**